KosherCocoa/Library/Core/Solar/KCNOAACalculator.m

```diff
@@ -0,0 +1,450 @@
+    return acos(cos(toRadians(zenith)) / (cos(latRad) * cos(sdRad)) - tan(latRad) * tan(sdRad)); // in radians
+}
+
+- (double)getSunHourAngleAtSunset:(double)lat solarDec:(double)solarDec zenith:(double)zenith {
+    double latRad = toRadians(lat);
+    double sdRad = toRadians(solarDec);
+    double hourAngle = acos(cos(toRadians(zenith)) / (cos(latRad) * cos(sdRad)) - tan(latRad) * tan(sdRad));
+    return -hourAngle; // in radians
+}
+
+- (double)getSunriseUTC:(double)julianDay latitude:(double)latitude longitude:(double)longitude zenith:(double)zenith {
+    double julianCenturies = [self getJulianCenturiesFromJulianDay:julianDay];
+    
+    // Find the time of solar noon at the location, and use that declination. This is better than start of the
+    // Julian day
+    double noonmin = [self getSolarNoonUTC:julianCenturies longitude:-longitude];
+    double tnoon = [self getJulianCenturiesFromJulianDay:julianDay + noonmin / 1440.0];
+    
+    // First pass to approximate sunrise (using solar noon)
+    double eqTime = [self getEquationOfTime:tnoon];
+    double solarDec = [self getSunDeclination:tnoon];
+    double hourAngle = [self getSunHourAngleAtSunrise:latitude solarDec:solarDec zenith:zenith];
+    
+    double delta = -longitude - toDegrees(hourAngle);
+    double timeDiff = 4 * delta;
+    double timeUTC = 720 + timeDiff - eqTime;
+    
+    // Second pass includes fractional Julian Day in gamma calc
+    double newt = [self getJulianCenturiesFromJulianDay:[self getJulianDayFromJulianCenturies:julianCenturies] + timeUTC/ 1440.0];
+    eqTime = [self getEquationOfTime:newt];
+    solarDec = [self getSunDeclination:newt];
+    hourAngle = [self getSunHourAngleAtSunrise:latitude solarDec:solarDec zenith:zenith];
+    
+    delta = -longitude - toDegrees(hourAngle);
+    timeDiff = 4 * delta;
+    timeUTC = 720 + timeDiff - eqTime;
+    return timeUTC;
+    }
+
+- (double)getSolarNoonUTC:(double)julianCenturies longitude:(double)longitude {
+    // First pass uses approximate solar noon to calculate equation of time
+    double tnoon = [self getJulianCenturiesFromJulianDay:[self getJulianDayFromJulianCenturies:julianCenturies] + longitude / 360.0];
+    double eqTime = [self getEquationOfTime:tnoon];
+    double solNoonUTC = 720 + (longitude * 4) - eqTime; // min
+    
+    double newt = [self getJulianCenturiesFromJulianDay:[self getJulianDayFromJulianCenturies:julianCenturies] - 0.5 + solNoonUTC / 1440.0];
+    eqTime = [self getEquationOfTime:newt];
+    return 720 + (longitude * 4) - eqTime; // min
+}
+
+
+- (double)getSunsetUTC:(double)julianDay latitude:(double)latitude longitude:(double)longitude zenith:(double)zenith {
+    double julianCenturies = [self getJulianCenturiesFromJulianDay:julianDay];
+    
+    // Find the time of solar noon at the location, and use that declination. This is better than start of the
+    // Julian day
+    double noonmin = [self getSolarNoonUTC:julianCenturies longitude:-longitude];
```

Comment:
nit:  Camel case `noonMin` and `tNoon` below.

Reply:
It was like that in the Java file, but I fixed it now. Either way, it looks better in camel case.